src/mpi/request/waitany.c

```diff
@@ -151,7 +151,7 @@ int MPI_Waitany(int count, MPI_Request array_of_requests[], int *indx, MPI_Statu
     MPIR_Request **request_ptrs = request_ptr_array;
     int i;
     int last_disabled_anysource = -1;
-    int first_nonnull = 0;
+    int first_nonnull = count;
```

Comment:
This should work, but I wonder if we could use `-1` to be more clear that `first_nonnull` does not yet have a meaningful value. `count` is close enough to a valid index that it could cause confusion.

Reply:
Thank! But still, in the case when all request_ptrs[i] are NULL, the expression 'first_nonnull = count' is true, but the expression 'first_nonnull = -1' in this case is not true. That is why initialization with 'first_nonnull = count' would be better.

Reply:
OK but here is another concernif all requests are NULL. On line 247 ```mpi_errno = MPID_Waitany(count - first_nonnull, &request_ptrs[first_nonnull], indx, status);``` we would be accessing beyond the bounds of the `request_ptrs` array. Some additional handling is needed going to be needed.

Reply:
@raffenet Please take into account that `&request_ptrs[first_nonnull]` is just a pointer which points beyond the bounds. Actually, there is no accessing beyond the bounds because `count - first_nonnull = 0` and the following call is safe:

`mpi_errno = MPID_Waitany(0, any_invalid_pointer, ...);`

Reply:
Well, I learned a new thing about undefined behavior in C. https://stackoverflow.com/questions/5341646/pointer-arithmetic-in-c-and-array-bounds. It looks like this is OK because we are only calculating the address of the first element beyond the bounds of the array.